src/sync/discovery.ts

```diff
@@ -1,14 +1,93 @@
-import { Subscription } from "../models";
 import { getRepositorySummary } from "./jobs";
 import enhanceOctokit from "../config/enhance-octokit";
 import { Application } from "probot";
-import { Repositories, SyncStatus } from "../models/subscription";
-import {LoggerWithTarget} from "probot/lib/wrap-logger";
+import Subscription, { Repositories, Repository, SyncStatus } from "../models/subscription";
+import { LoggerWithTarget } from "probot/lib/wrap-logger";
 import { sqsQueues } from "../sqs/queues";
+import GitHubClient from "../github/client/github-client";
+import { getCloudInstallationId } from "../github/client/installation-id";
+import { DiscoveryMessagePayload } from "../sqs/discovery";
 
-export const DISCOVERY_LOGGER_NAME = "sync.discovery";
+/*
+* Mapping the response data into a map by repo.id as required by subscription.updateSyncState.
+*/
+const mapRepositories = (repositories: Repository[]): Repositories => {
+	return repositories.reduce((obj, repo) => {
+		obj[repo.id] = { repository: getRepositorySummary(repo) };
+		return obj;
+	}, {});
+};
+
+/*
+* Reset the sync count to zero.
+*/
+const resetSyncedReposCount = async (subscription: Subscription): Promise<void> => {
+	await subscription.updateSyncState({ numberOfSyncedRepos: 0 });
+};
+
+/*
+* Update the sync status of a batch of repos.
+*/
+const updateSyncState = async (subscription: Subscription, repositories: Repository[]): Promise<void> => {
+	const repos = mapRepositories(repositories);
+	await subscription.updateSyncState({
+		repos
+	});
+};
+
+/*
+* Checking the header atttribute link(type string[]) for a entry that contains "rel=next", this indicates more pages available.
+*/
+const checkHeaderForNextPage = (headers): boolean => {
+	const links = headers?.link?.split(",");
+	const regex = /(rel="next")/g;
+	const hasNextPage = links?.find(elem => elem.match(regex));
+	return hasNextPage;
+};
+
+/*
+* Continuosuly call the GitHub repo to fetch a page of repositories at a time and update there sync status until no more pages.
+*/
+const syncRepositories = async (github, subscription: Subscription, logger: LoggerWithTarget): Promise<void> => {
+	let page = 1;
```

Comment:
Page = 1, page zero is invalid on the GitHub side so it defaults to page 1, this would cause our counter to be behind by 1 and receive page 1 twice.... therefore starting at page one.

Reply:
I don't like the idea that the implementation of the API is bleeding out to this, same with the  `checkHeaderForNextPage` function.  Can we maybe add the `hasNextPage` and the next page number be returned as part of the `getRepositories` function?

Another possibility that I think might be the best way, is having a `pagination` function in the github client which knows about pages and `hasNextPage` implementation details and essentially does the same while loop and then just calls the callback when a page arrives.  Something like:

```
await github.paginate('getRepositoriesPage', async (data) => { /* do the thing */ });
```
It might be outside the scope of this PR, but would be extremely useful for the future.

Reply:
Do we even need to supply the page number for the first page?

Also I guess the best way will be to not reinvent the wheel and see how it is done in Octokit

Reply:
We can omit the page and it will return page one by default, in the link header it contains the full url for the next page.... i think(outside of this PR) when we to at implementing a more complete solution like `await github.paginate('getRepositoriesPage', async (data) => { /* do the thing */ })` we can try use that then/ see what Octokit does? I imagine this is going to be re-used a fair bit, so worth doing it proper.